blockchain/utxocache.go

```diff
@@ -235,6 +235,125 @@ type utxoBatcher interface {
+	// should contain in normal circumstances.
+	maxTotalMemoryUsage uint64
+}
+
+// length returns the length of all the maps in the map slice added together.
+func (ms *mapSlice) length() int {
+	var l int
+	for _, m := range ms.maps {
+		l += len(m)
+	}
+
+	return l
+}
+
+// size returns the size of all the maps in the map slice added together.
+func (ms *mapSlice) size() int {
+	var size int
+	for _, num := range ms.maxEntries {
+		size += CalculateRoughMapSize(num, bucketSize)
+	}
+
+	return size
+}
+
+// get looks for the outpoint in all the maps in the map slice and returns
+// the entry.  nil and false is returned if the outpoint is not found.
+func (ms *mapSlice) get(op wire.OutPoint) (*UtxoEntry, bool) {
+	var entry *UtxoEntry
+	var found bool
+
+	for _, m := range ms.maps {
+		entry, found = m[op]
+		if found {
+			break
+		}
+	}
+
+	return entry, found
+}
+
+// put puts the outpoint and the entry into one of the maps in the map slice.  If the
+// existing maps are all full, it will allocate a new map based on how much memory we
+// have left over.  Leftover memory is calculated as:
+// maxTotalMemoryUsage - (totalEntryMemory + mapSlice.size())
+func (ms *mapSlice) put(op wire.OutPoint, entry *UtxoEntry, totalEntryMemory uint64) {
+	var success bool
+	for i, maxNum := range ms.maxEntries {
+		m := ms.maps[i]
+		_, found := m[op]
+		if found {
+			// If the key is found, overwrite it.
+			m[op] = entry
+			success = true
+			continue
+		}
+		if len(m) >= maxNum {
+			// Don't try to insert if the map already at max since
+			// that'll force the map to allocate double the memory it's
+			// currently taking up.
+			continue
+		}
+
+		m[op] = entry
+		success = true
+	}
+
+	if !success {
+		ms.makeNewMap(totalEntryMemory)
+		m := ms.maps[len(ms.maps)-1]
+		m[op] = entry
+	}
+}
+
+// delete attempts to delete the given outpoint in all of the maps. No-op if the
+// outpoint doesn't exist.
+func (ms *mapSlice) delete(op wire.OutPoint) {
+	for i := 0; i < len(ms.maps); i++ {
+		delete(ms.maps[i], op)
+	}
+}
+
+// makeNewMap makes and appends the new map into the map slice.
+func (ms *mapSlice) makeNewMap(totalEntryMemory uint64) {
+	// Get the size of the leftover memory.
+	memSize := ms.maxTotalMemoryUsage - totalEntryMemory
+	for _, maxNum := range ms.maxEntries {
+		memSize -= uint64(CalculateRoughMapSize(maxNum, bucketSize))
+	}
+
+	// Get a new map that's sized to house inside the leftover memory.
+	numMaxElements := CalculateMinEntries(int(memSize), bucketSize+avgEntrySize)
+	numMaxElements -= 1
+	ms.maxEntries = append(ms.maxEntries, numMaxElements)
+	ms.maps = append(ms.maps, make(map[wire.OutPoint]*UtxoEntry, numMaxElements))
+}
+
+// deleteMaps deletes all maps except for the first one which should be the biggest.
+func (ms *mapSlice) deleteMaps() {
+	// The rest of the map is now available to be garbage collected.
+	m := ms.maps[0]
+	ms.maps = []map[wire.OutPoint]*UtxoEntry{m}
```

Comment:
Shouldn't this once again pre-allocate? Otherwise, it needs to go through the successive doubling as new items are added.

Reply:
The oldest map/largest is kept and the smaller ones are left out so they can be gc-ed.

The current algorithm will take a look at the leftover memory and allocate a map that will fit inside of that. My hunch is that since the entry sizes can vary, it's better to let the smaller map be gc-ed so the second map size can be dynamic based on the needs at that moment.